File: common/src/main/kotlin/net/impleri/playerskills/skills/tiered/TieredSkillType.kt
package net.impleri.playerskills.skills.tiered

import net.impleri.playerskills.PlayerSkills
import net.impleri.playerskills.api.Skill
import net.impleri.playerskills.api.SkillType
import net.impleri.playerskills.utils.SkillResourceLocation
import net.minecraft.resources.ResourceLocation

class TieredSkillType : SkillType<String>() {
  override val name: ResourceLocation
    get() = NAME

  override fun castToString(value: String?): String? {
    return value
  }

  public override fun castFromString(value: String?): String? {
    return value
  }

  override fun can(skill: Skill<String>, expectedValue: String?): Boolean {
    val givenValue = getIndexFor(skill.value, skill.options) ?: 0
    val testValue = getIndexFor(expectedValue, skill.options) ?: 1

    PlayerSkills.LOGGER.debug("Checking if player can ${skill.name} (is ${skill.value}->$givenValue >= $testValue<-$expectedValue)")

    return givenValue >= testValue
  }

  override fun getPrevValue(skill: Skill<String>, min: String?, max: String?): String {
    val currentValue = getIndexFor(skill.value, skill.options) ?: -1
    val maxIndex = getIndexFor(max, skill.options) ?: skill.options.size
    val minIndex = getIndexFor(min, skill.options) ?: 0

    val prevVal = (currentValue - 1).coerceAtLeast(minIndex).coerceAtMost(maxIndex)

    return skill.options[prevVal]
  }

  override fun getNextValue(skill: Skill<String>, min: String?, max: String?): String {
    val currentValue = getIndexFor(skill.value, skill.options) ?: -1
    val maxIndex = getIndexFor(max, skill.options) ?: skill.options.size
    val minIndex = getIndexFor(min, skill.options) ?: 0

    val nextVal = (currentValue + 1).coerceAtLeast(minIndex).coerceAtMost(maxIndex)

    return skill.options[nextVal]
  }

  private fun getIndexFor(value: String?, options: List<String>): Int? {
    value ?: return null
    val index = options.indexOf(value)

    return if (index == -1) null else index
  }

  companion object {
    var NAME = SkillResourceLocation.of("tiered")
  }
}

File: common/src/main/scala/net/impleri/playerskills/skills/tiered/TieredSkillType.scala
package net.impleri.playerskills.skills.tiered

import net.impleri.playerskills.api.skills.Skill
import net.impleri.playerskills.api.skills.SkillOps
import net.impleri.playerskills.api.skills.SkillType
import net.impleri.playerskills.utils.PlayerSkillsLogger
import net.impleri.slab.logging.Logger
import net.impleri.slab.resources.ResourceLocation

import scala.util.chaining.scalaUtilChainingOps
import scala.util.Try

case class TieredSkillType(
  override val skillOps: SkillOps = Skill(),
  private val logger: Logger = PlayerSkillsLogger.SKILLS,
) extends SkillType[String] {
  override val name: ResourceLocation = TieredSkillType.NAME

  override def castToString(value: String): Option[String] = Option(value)

  override def castFromString(value: String): Option[String] = Option(value)

  private def indexOf(
    skill: Skill[String],
    target: Option[String] = None,
  ): Option[Int] =
    target
      .map(skill.options.indexOf)
      .filter(_ >= 0)
      .filter(_ < skill.options.size)

  private def compare(
    skill: Skill[String],
    threshold: Option[String],
  ): Boolean =
    (indexOf(skill, skill.value), indexOf(skill, threshold)) match {
      case (Some(v), Some(t)) => v >= t
      case (Some(_), _)       => true
      case _                  => false
    }

  override def can(
    skill: Skill[String],
    threshold: Option[String] = None,
  ): Boolean =
    compare(skill, threshold)
      .tap(
        logger.debugP(c =>
          s"Checking if player can ${skill.name} (is $threshold above ${skill.value}? $c)",
        ),
      )

  private def floor(
    skill: Skill[String],
    value: Int,
    min: Option[String],
  ): Int = indexOf(skill, min).fold(value)(value max _)

  private def ceil(skill: Skill[String], value: Int, max: Option[String]): Int =
    indexOf(skill, max).fold(value)(value min _)

  override def getPrevValue(
    skill: Skill[String],
    min: Option[String],
    max: Option[String],
  ): Option[String] =
    for {
      next <- indexOf(skill, skill.value).map(_ - 1)
      floored = floor(skill, next, min)
      ceiling = ceil(skill, floored, max)
      nextString <- Try(skill.options.apply(ceiling)).toOption
    } yield nextString

  override def getNextValue(
    skill: Skill[String],
    min: Option[String],
    max: Option[String],
  ): Option[String] =
    for {
      next <- indexOf(skill, skill.value).map(_ + 1)
      floored = floor(skill, next, min)
      ceiling = ceil(skill, floored, max)
      nextString <- Try(skill.options.apply(ceiling)).toOption
    } yield nextString
}

object TieredSkillType {
  val NAME: ResourceLocation = ResourceLocation("tiered").get
}
